### DISCUSSION TITLE
Extract web::Json from application middleware proves problematic

### DISCUSSION THREAD

**Discussion by Ducolnd** [actix/actix-web#3436] [Q&A] [answered]
Hi. I have written a piece of middleware in which I need to extract `web::Json<T>` which works as expected as follows:

```rust
 fn call(&self, mut req: ServiceRequest) -> Self::Future {
        let service = self.service.clone();
        let guard = self.inner.clone(); 

        Box::pin(async move {
            let extracted_json = req.extract::<web::Json<Club>>().await;
            println!("{:?}", extracted_json);
            
           // Downstream code and call service
           // ......
        })
    }

```

However, when trying to extract json twice in a row like this, it fails:

```rust
let extracted_json = req.extract::<web::Json<Club>>().await;
println!("Extract 1 {:?}", extracted_json);

let extracted_json2 = req.extract::<web::Json<Club>>().await;
println!("Extract 2 {:?}", extracted_json2);
```

Extract 1 works as expected, whereas Extract 2 fails with `Json deserialize error: EOF while parsing a value at line 1 column 0`. The problem I'm trying to illustrate here is that this also fails when I extract json the first time in the middleware (which works as expected) and a second time somewhere downstream, i.e. in a handler like this, which fails:

```rust
async fn create(club: web::Json<Club>) -> WebResult<HttpResponse> {
    let club = club.into_inner();
    // More code
    // .........
}
```

This is extremely annoying for my application because this means I cannot extract Json in the middleware without also breaking the program somewhere further on (in the handler). Why is this? How do I resolve this? Should I open an issue addressing this?

**Comment by robjtede** [comment, marked as answer]
Extractors that consume the request body can only do so once, or else we'd be potentially needing to buffer huge payloads in-memory.

[One of the examples](https://github.com/actix/examples/blob/87d90cc0d000e8240f6d0744ccf53365bc7798b7/middleware/encrypted-payloads/src/main.rs#L116-L123) shows a solution to this, which is to put a body back into the request after consuming it, allowing inner middleware and handlers to consume it again.

**Reply by Ducolnd** [reply]
Thank you a lot! I wasn't aware of this
